Document proposed v4 serialization format (#505)
dmlc/treelite: docs/serialization/v4.rst

================================
Treelite Serialization Format v4
================================

The v4 serialization format was designed with the following goals in mind:

* First-class support for multi-target models
* Support "boosting from the average" in scikit-learn, where a simple base estimator is fitted from the class label distribution (or the average label, for regression) and is used as the initial learner in the ensemble model.
* Use integer types with defined widths (so ``int32_t`` instead of ``int``)

We first define a set of enum types to be used in the serialization format.

* ``TypeInfo``: underlying type ``uint8_t``. Indicates the data type of another field

  - ``kInvalid`` (0)
  - ``kUInt32``  (1)
  - ``kFloat32`` (2)
  - ``kFloat64`` (3)

* ``TaskType``: underlying type ``uint8_t``. Indicates the type of the learning task.

  - ``kBinaryClf`` (0): binary classifier
  - ``kRegressor`` (1): regressor
  - ``kLearningToRank`` (2): learning-to-rank
  - ``kMultiClfGrovePerClass`` (3): multi-class classifier with grove-per-class

>   - ``kMultiClfProbDistLeaf`` (4): multi-class classifier with vector leaf outputs producing probability distribution

Author: ```suggestion
  - ``kMultiClfProbDistLeaf`` (4): multi-class classifier with vector leaf outputs producing probability distribution
  - ``kIsolationForest`` (5): isolation forest
```

   * Task type: single ``uint8_t`` scalar representing enum ``TaskType``.
   * Whether to average tree outputs: single ``bool`` scalar.
   * Task parameters

     - ``num_target``: single ``uint32_t`` scalar. Number of targets in the model. ``num_target > 1`` indicates a multi-target models.
     - ``num_class``: an array of ``uint32_t`` with length ``num_target``.
     - ``leaf_vector_shape``: an array of ``uint32_t`` with length 2. The first dimension is always ``num_target``. The second dimension is either 1 or ``max(num_class)``.
     - ``grove_per_class``: single ``bool`` scalar

   * Model parameters (``ModelParam``) a structure with following fields

     - ``pred_transform``: 256-character long ``char`` array. Stores a human-readable name of the transformation function that's applied to prediction outputs. The unused elements in the array should be padded with null characters (``\0``).
     - ``sigmoid_alpha``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="sigmoid"``.
     - ``ratio_c``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="exponential_standard_ratio"``.

>    * ``base_scores``: an array of ``float``. This vector is expected to have length ``num_target``.

Author: ```suggestion
   * ``base_scores``: an array of ``LeafOutputType``. This vector is expected to have length ``num_target``.
```

   * Task type: single ``uint8_t`` scalar representing enum ``TaskType``.
   * Whether to average tree outputs: single ``bool`` scalar.
   * Task parameters

     - ``num_target``: single ``uint32_t`` scalar. Number of targets in the model. ``num_target > 1`` indicates a multi-target models.
     - ``num_class``: an array of ``uint32_t`` with length ``num_target``.
     - ``leaf_vector_shape``: an array of ``uint32_t`` with length 2. The first dimension is always ``num_target``. The second dimension is either 1 or ``max(num_class)``.
     - ``grove_per_class``: single ``bool`` scalar

   * Model parameters (``ModelParam``) a structure with following fields

     - ``pred_transform``: 256-character long ``char`` array. Stores a human-readable name of the transformation function that's applied to prediction outputs. The unused elements in the array should be padded with null characters (``\0``).
     - ``sigmoid_alpha``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="sigmoid"``.
     - ``ratio_c``: single ``float`` scalar. This model parameter is relevant when ``pred_transform="exponential_standard_ratio"``.

>    * ``base_scores``: an array of ``LeafOutputType``. This vector is expected to have length ``num_target``.

Author: ```suggestion
   * ``base_scores``: an array of ``LeafOutputType``. This vector is expected to have length ``num_target * max(num_class)``. The elements will be laid out in the row-major layout.
```

     - ``cleft_``: single ``int32_t`` scalar. Indicates the ID of the left child node. Set to ``-1`` to indicate the lack of the left child.
     - ``cright_``: single ``int32_t`` scalar. Indicates the ID of the right child node. Set to ``-1`` to indicate the lack of the right child.
     - ``sindex_``: single ``uint32_t`` scalar. This field gives both the feature ID used in the current test node (``split_index``), as well as the default direction for the missing value (``default_left``). Set this value by computing ``split_index | (default_left ? (1U << 31U) : 0)``.
     - ``info_``: a union type containing ``leaf_value`` (of type ``LeafOutputType``) and ``threshold`` (of type ``ThresholdType``). To set this field, determine whether the node is a leaf node or an internal test node. Use ``leaf_value`` for leaf nodes; use ``threshold`` for internal test nodes.
     - ``comp_``: single ``int8_t`` scalar representing enum ``Operator``.
     - ``categories_list_right_child_``: single ``bool`` scalar.

   * Leaf vectors

     - Content (``leaf_vector_``): an array of ``LeafOutputType``. This array stores the leaf vectors for all nodes, such that

>        the sub-array ``leaf_vector_[leaf_vector_begin[i]_:leaf_vector_end_[i]]`` yields the leaf vector for the i-th node.

Author: ```suggestion
       the sub-array ``leaf_vector_[leaf_vector_begin[i]_:leaf_vector_end_[i]]`` yields the leaf vector for the i-th node.
       The leaf vector uses the row-major layout to store a 2D array.
```

     - ``sum_hess_present_``: an array of ``bool``.  ``sum_hess_present_[i]`` indicates whether ``sum_hess_[i]`` is available.
     - ``gain_``: an array of ``double``.  ``gain_[i]`` indicates the change in the loss function that is attributed to the particular split at node ``i``.
     - ``gain_present_``: an array of ``bool``. ``gain_present_[i]`` indicates whether ``gain_[i]`` is present.

   * Extension slot 2: Per-tree optional fields. This field is currently not used.

     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.

   * Extension slot 3: Per-node optional fields. This field is currently not used.

     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.

#. Tree 1: Use the same set of fields as Tree 0.
#. Other trees ...

.. note:: Caveat for multi-target, multi-class classifiers

   When the number of classes are different for targets, we use the larget number of

>    classes (``max_num_class``) and use it to shape the leaf vector. The leaf vector

Author: ```suggestion
   classes (``max_num_class``) and use it to shape the leaf vector (and ``base_scores``). The leaf vector
```

     - ``sum_hess_present_``: an array of ``bool``.  ``sum_hess_present_[i]`` indicates whether ``sum_hess_[i]`` is available.
     - ``gain_``: an array of ``double``.  ``gain_[i]`` indicates the change in the loss function that is attributed to the particular split at node ``i``.
     - ``gain_present_``: an array of ``bool``. ``gain_present_[i]`` indicates whether ``gain_[i]`` is present.

   * Extension slot 2: Per-tree optional fields. This field is currently not used.

     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.

   * Extension slot 3: Per-node optional fields. This field is currently not used.

     - Number of fields: single ``int32_t`` scalar. Set this value to ``0``, to indicate the lack of optional fields.

#. Tree 1: Use the same set of fields as Tree 0.
#. Other trees ...

.. note:: Caveat for multi-target, multi-class classifiers

   When the number of classes are different for targets, we use the larget number of

>    classes (``max_num_class``) and use it to shape the leaf vector (and ``base_scores``). The leaf vector

Author: ```suggestion
   When the number of classes are different for targets, we use the larget number of
   classes (``max_num_class``) to shape the leaf vector (and ``base_scores``). The leaf vector
```

>    will have shape ``(num_target, max_num_class)``. This heuristic has the following

Author: ```suggestion
   will have shape ``(num_target, max_num_class)``, with extra elements padded with ``0``. This heuristic has the following
```
